llvm/include/llvm/IR/InstrTypes.h

```diff
@@ -1033,6 +1033,28 @@ class CmpInst : public Instruction {
   }
 };
 
+/// An icmp instruction, which can be marked as "samesign", indicating that the
+/// two operands have the same sign. This means that we can convert "slt/ult"
+/// to "ult", which enables more optimizations.
+class PossiblySameSignInst : public CmpInst {
```

Comment:
Why is the base class `CmpInst` and not `ÌCmpInst`?

Reply:
I'd prefer to add hasSameSign/setSameSame directly to ICmpInst, instead of having the PossiblySameSignInst wrapper.

Reply:
> Why is the base class `CmpInst` and not `ÌCmpInst`?

Oops, I missed this!

Reply:
> I'd prefer to add hasSameSign/setSameSame directly to ICmpInst, instead of having the PossiblySameSignInst wrapper.

@nikic  Is there a specific reason for this?

Reply:
ICmpInst is the only instruction that supports the flag, so adding the methods directly there will reduce the amount of casting that has to be done.

Similar to how nuw/nsw are directly set on TruncInst.